### DISCUSSION TITLE
Localization

### DISCUSSION THREAD

**Discussion by JanOlsmar** [oqtane/oqtane.framework#2273] [Q&A] [answered]
I am testing and building moduls in Oqtane it is fun. Localization and globalization is never fun :)
For the moment I just need Oqtane and my moduls to work  for, in my case , Swedish culture. Decimal comma as seperator and right format for other data.  As I understand it Oqtane shows all in culture en.
To get it work for other culture you need to do a languge translation package and install it.
Is this right? It must be some better way to get this to work. I tryed to change the appsetting but ths had no impact.
Please give me some advice.

**Comment by hishamco** [comment]
> Is this right?

Either you create a language pack and install it, or add the required resx files to your module

> It must be some better way to get this to work

Your suggestion

There are some changes happen after put my hands on it, hope @sbwalker clarify what is the state of it from module development perspective

**Comment by JanOlsmar** [comment]
Thanks for answer but it seems I was unclear.  To add resx file to my module does not solve my problem. 
In fact I don't care about the translation of text. I need Oqtane have sv-SE as culture.
I need it to get right format.

**Reply by hishamco** [reply]
The localization should respect any culture if it's supported, Oqtane already allowed you add the supported cultures, so there's no issue. But this should localize your app, number, currency format as well. In case you need to localize the strings in your app, you must add the resx files

**Comment by sbwalker** [comment, marked as answer]
@JanOlsmar I think I understand what you are asking and I can provide some context around the current implementation. When you choose the Add Language option for a Site the following UI is displayed. Originally the dropdown contained a list of cultures that were specified in appsettings.json. The problem with this approach was that the list of cultures was static, not modifiable without access to the server, and did not reflect the list of localized translations which were actually installed. So the logic was modified so that the list of cultures is based on the list of translations that are actually installed (ie. that exist in the satellite subfolders of /bin which contain *.resources.dll files ). 

![image](https://user-images.githubusercontent.com/4840590/177826058-cf564e63-6716-4a67-b6ed-ea1df6482bc6.png)

The list of languages which are enabled for a Site are available in the Language Picker - which sets the culture cookie for the user session:

![image](https://user-images.githubusercontent.com/4840590/177826919-66e33a5b-e6dd-4513-b2b6-ce46f0625fc4.png)

However you would like to be able to add a culture to your site - without a corresponding translation. The scenario is that you would like all of the static content displayed in English however you would like globalized information such as dates, currency, etc... to be displayed using a different culture format. Is this correct?

**Comment by JanOlsmar** [comment]
Yes thats correct.
I added a Settings.sv_SE.resx file to the HTML module an recompiled my developement installation. So got it to work for that purpose.
But I think it would be nice to add a language without any translation.

**Comment by sbwalker** [comment]
Thinking more about this... ASP.NET has 2 distinct settings for culture:

Thread.CurrentThread.CurrentCulture - specifies the globalization options for date, currency, etc...
Thread.CurrentThread.CurrentUICulture - specifies the localization resources to use

Currently Oqtane only supports a single culture specification... and it sets both the CurrentCulture and CurrentUICulture based on this specification  - which means that you can not specify a CurrentCulture of Swedish and a CurrentUICulture of English (which is essentially what you are asking for).

Many applications allow a user to specify their preferred Language (CurrentUICulture) and Regional Format (CurrentCulture). This is usually entered in their user profile. It would not be complicated to add this to Oqtane. However, it does make me wonder what we would do with the existing language picker component...

**Reply by JanOlsmar** [reply]
It works fine with both Culture and UICulture in Swedish. As long as if not swedish translation exist it takes the default english.
This is the best solution. It make it possible to translate modules that needs to be translated.
Swedish Culture and English UI culture is a bad combination. It make it hard to translate if needed.

**Reply by JanOlsmar** [reply]
And if the bowser language setting forced the culture to be set.

**Comment by sbwalker** [comment]
@JanOlsmar #2292 should resolve the issue you mentioned in this thread. It will allow an Administrator to add any language in Language Management - regardless of whether a translation exists or not. If a language does have a translation installed, it will now display the version of the translation.

**Reply by JanOlsmar** [reply]
Perfect.
I have follow up question regarding translation of modules. I can't get it work. Framework is ok but my modules does not pick up the translations. Is this not yet implemented in Oqtane or do I do something wrong. I use the  Module creator an this has prepared for it so I asume it should work.

**Reply by sbwalker** [reply]
It definitely works for modules - maybe @leigh-pointer can offer some assistance. The main thing which seems to cause issues for developers is the namespace specification.

**Reply by JanOlsmar** [reply]
Ok thanks. I will look deeper in to it.
